stl/inc/ranges

```diff
@@ -1433,6 +1433,137 @@ namespace ranges {
         inline constexpr _Istream_fn<_Ty> istream;
     } // namespace views
 
+    template <class _Derived>
+        requires is_class_v<_Derived> && same_as<_Derived, remove_cv_t<_Derived>>
+    class range_adaptor_closure;
+
+    namespace _StdPipe {
+        template <class _Ty>
+        void _Derived_from_range_adaptor_closure_impl(const volatile range_adaptor_closure<_Ty>&);
+
+        template <class _Ty>
+        concept _Derived_from_range_adaptor_closure = requires(const volatile _Ty& _Obj) {
+            _Derived_from_range_adaptor_closure_impl(_Obj);
+        };
+
+        template <class _Ty>
+        concept _Is_range_adaptor_common = derived_from<remove_cvref_t<_Ty>,
+            range_adaptor_closure<remove_cvref_t<_Ty>>> && _Derived_from_range_adaptor_closure<_Ty> && !range<_Ty>;
+
+        template <class _Rng, class _Ty>
+        concept _Is_invocable_range_adaptor_closure =
+            range<_Rng> && invocable<_Ty, _Rng> && _Is_range_adaptor_common<_Ty>;
+
+        template <class _Left, class _Right>
+        concept _Are_composable_range_adaptor_closures =
+            _Is_range_adaptor_common<_Left> && _Is_range_adaptor_common<_Right> && _Pipe::_Can_compose<_Left, _Right>;
+
+        template <class _Left, class _Right>
+        struct _StdPipeline;
+    } // namespace _StdPipe
+
+    template <class _Derived>
+        requires is_class_v<_Derived> && same_as<_Derived, remove_cv_t<_Derived>>
+    class range_adaptor_closure {
+    public:
+        template <class _Other>
+            requires _StdPipe::_Are_composable_range_adaptor_closures<_Derived&, _Other>
+        constexpr auto operator|(_Other&& __r) & noexcept(

```

Comment:
We conventionally use `_Ugly` names and not `__ugly` ones.

The one difference is inside `requires` clauses, there the arguments are `__ugly` but everything else is not. Also we try to avoid single letter names, so `_Left`, `_Right`, `_Range` etc